h/views/oidc.py

```diff
@@ -164,18 +167,18 @@ def redirect(self):
             # We received an invalid or unexpected redirect from ORCID.
             raise
 
-        decoded_state = _decode_oauth2_state_param(
-            validated_params["state"], state_signing_key
+        decoded_state = self._jwt_service.decode_symmetric(
+            validated_params["state"],
+            audience=JWTAudiences.OIDC_REDIRECT_ORCID,
+            payload_class=OIDCState,
         )
 
-        action_str = decoded_state["action"]
-
-        if action_str == "connect" and not self._request.is_authenticated:
+        if decoded_state.action == "connect" and not self._request.is_authenticated:
             # You must be logged in to connect an identity to your existing
             # Hypothesis account.
             raise HTTPForbidden
 
-        if action_str == "login" and self._request.is_authenticated:
+        if decoded_state.action == "login" and self._request.is_authenticated:
```

Comment:
I found an edge case here:

1. Open two tabs while logged out
2. Click "Connect with ORCID" in one tab. When redirected to ORCID, don't sign in yet.
3. Repeat in the other tab
4. Complete login in first tab
5. Complete login in second tab. You'll get "Page not found" after redirecting back to Hypothesis

If you click "Connect with ORCID" on the login form in the second tab after logging in on the first tab, you'll be immediately redirected to your user search page.

Reply:
Good find. What's happening is that the second tab overwrites the first tab's OAuth 2 `state` in the browser session. The code then tries to redirect to `/account/settings` with a "Received invalid redirect from ORCID" flash message, which would make sense if you were already logged in to Hypothesis and were trying to connect your ORCID iD to your existing Hypothesis account, but if you're trying to log in or sign up the `/account/settings` page 404s because you're not logged in yet. I've created an issue to fix this: https://github.com/orgs/hypothesis/projects/158?pane=issue&itemId=118496301